### DISCUSSION TITLE
Unlocking creativity within the constraints of corporate development

### DISCUSSION THREAD

**Discussion by geeprime** [community/community#141819] [Other Feature Feedback, Questions, & Ideas]
Universe Session Track: Developer Experience [DEV1459L](https://reg.githubuniverse.com/flow/github/universe24/attendee-portal/page/sessioncatalog/session/1715291080280001Nea7)

> Enterprises are tightly regulated and necessary controls limit what employees can do, often to the detriment of developers. A developer's ability to innovate and experiment is stymied by privileged access to corporate-issued devices and developer tools that are pre-configured and deployed by central IT. Join this discussion facilitated by Jim Richer, distinguished engineer, at TD Bank, to take a look at what these controls are and how to enforce them—all while still championing creativity. 

I'm excited to learn about what the developer experience is like for all software engineers, and how it may be impacted by corporate controls implemented by your respective control partners (and why it feels like someone called "the fun police" when new controls are implemented). These controls are important, as they exist to protect the customers of their respective organizations. They cannot be ignored, so lets discuss what approaches we can take to foster creativity within those constraints!

I would love to learn:

1. What industry are you working in?
1. What are the controls that impact you, and why do they matter?
1. What do these control implementations look like, and how do they get in the way?
1. How can we respect the spirit of the control, and develop solutions that encourage creativity within those controls?
1. When necessary, how can we find more creative ways to re-imagine control implementations that continue to protect the corporation and its customers, while ensuring developers have more freedom to focus on delivering more great features to those same customers?

I'm excited to hear what you all have to say! If you happen to attend GitHub Universe 2024, please swing by for some in-person feedback!

**Comment by nzxwang** [comment]
Hi @geeprime thanks for hosting the discussion session. It was interesting to learn about the different ways in which control partners at various companies implement controls and the effects that they have on the developer experience.

I work in the gaming industry at Riot Games but specifically on a team that owns an observability pipeline that collects all internal telemetry before forwarding to Datadog. In some ways, its just a b2b integration. There are many controls that a corporation like Riot Games enforces on engineers like myself but my team is also a control partner to the rest of Riot Games. For example, we set limits on the volume of logs and cardinality of metrics that a service can send in an effort to force our customers to think about the associated costs. We've received a lot of critical feedback about how these controls limit our customers from ultimately serving our players so we've implemented simple policies for when services can have exceptions as well as opt-in alternatives for sending telemetry to different data stores with less limits.

**Reply by geeprime** [reply]
It's definitely interesting when we see control implementations that control the costs of an organization, a worthy cause.

Sometimes, "showing engineers the bill" (with an internal chargeback mechanism, for example) is enough education to help them determine the value of their telemetry, and if sending more is worth the cost (or if they need to prune some unused data). That can often avoid the requirement for those exception paths as well.

How do you feel cost controls like this should be implemented? Now that those exception paths exist, what sort of feedback are you receiving from your consumers on their ability to serve their players? Has their experience as developers improved?

**Reply by donert** [reply]
Engineering is about design within constraints, one of which is money. As is oft said, an engineer can build for $1 what any damn fool can build for $5.

**Comment by Gesronet** [comment]
Hi Jim,

Thank you for initiating this important discussion. I completely agree that while corporate controls are essential for protecting our customers and organizations, they can sometimes hinder developers’ creativity and innovation.

I work as a freelancer, and I've experienced firsthand how certain controls impact my workflow. For instance, working with clients who have strict security protocols can limit my access to necessary tools and environments, making it challenging to deliver projects efficiently. These implementations can often feel restrictive, hindering my ability to explore new ideas or technologies.

I'm particularly interested in exploring how we can balance compliance and creativity. I believe there are opportunities to respect the intent of these controls while still fostering an environment where developers can thrive. Perhaps we can brainstorm ways to streamline processes or leverage tools that provide the necessary safeguards without stifling innovation.

I look forward to hearing insights from others and discussing potential solutions. If I’m at GitHub Universe 2024, I’ll stop by to share more thoughts in person!

Best,

Samuel Otomewo

**Comment by Gasila5** [comment]
0xaf4519c35952619E89C1ac2dE521c21Cbf891346
